python-client/giskard/slicing/text_slicer.py

```diff
@@ -85,44 +80,103 @@ def find_metadata_slices(self, feature, target):
+
+        # Get top tokens by TF-IDF
+        order = np.argsort(tfidf.max(axis=0).toarray().squeeze())[::-1]
+        top_tokens = vectorizer.get_feature_names_out()[order[:max_tokens]]
+
+        return list(top_tokens)
+
+    def _get_high_loss_tokens(self, feature, target, max_tokens=1000):
+        from scipy import stats
+
+        vectorizer = _make_vectorizer(self.dataset.df[feature], tfidf=True)
+        tfidf = vectorizer.transform(self.dataset.df[feature])
+
+        lrank = self.dataset.df[target].rank(pct=True)
+
+        # If the vocabulary is too large, prefilter top tokens
+        # @TODO: check this
+        vocab_size = tfidf.shape[1]
+        if vocab_size > max_tokens * 10:
+            token_ns = np.argpartition(tfidf.max(axis=0).toarray().squeeze(), max_tokens * 10 - 1)[: max_tokens * 10]
+        else:
+            token_ns = np.arange(vocab_size)
+
+        # Find tokens which are most correlated with loss
+        rank_corrs = np.asarray([stats.spearmanr(tfidf[:, n].toarray().squeeze(), lrank)[0] for n in token_ns])
+        token_idx = token_ns[rank_corrs.argsort()[:max_tokens]]
 
-    def _get_top_tokens(self, feature_data, n=30):
-        from sklearn.feature_extraction.text import TfidfVectorizer
+        return list(vectorizer.get_feature_names_out()[token_idx])
 
-        raw_stopwords = sw_en + sw_fr
+    def _get_deviant_tokens(self, feature, target, max_tokens=100):
+        from scipy import stats
 
-        tokenizer = TfidfVectorizer().build_tokenizer()
-        tokenized_stopwords = sum([tokenizer(stop_word) for stop_word in raw_stopwords], [])
+        vectorizer = _make_vectorizer(self.dataset.df[feature], tfidf=False, binary=True)
+        X = vectorizer.transform(self.dataset.df[feature])
 
-        text_data = feature_data.values.astype("U")
-        vectorizer = TfidfVectorizer(stop_words=tokenized_stopwords)
-        tfidf = vectorizer.fit_transform(text_data)
+        critical_target = self.dataset.df[target].quantile(0.75)
+        y = self.dataset.df[target] > critical_target
+        Y = np.stack((1 - y, y), axis=1)
 
-        vocab = vectorizer.vocabulary_
-        inv_vocab = {v: k for k, v in vocab.items()}
+        counts = X.T @ Y
+        totals = Y.sum(axis=0)
+        remainders = counts - totals
+        tokens = vectorizer.get_feature_names_out()
 
-        # Compute the global TF-IDF score for each word.
-        global_tfidf = np.asarray(tfidf.mean(axis=0)).squeeze()
-        sorted_global_tfidf_indices = global_tfidf.argsort()[::-1]
+        mask = (counts.max(axis=-1) > 5) & (counts.min(axis=-1) > 0) & (remainders.min(axis=-1) > 0)
 
-        # Get the top n words sorted by global TF-IDF score.
-        top_words = [inv_vocab[idx] for idx in sorted_global_tfidf_indices[:n]]
+        _data = []
+        for token, token_counts, token_remainders in zip(tokens[mask], counts[mask], remainders[mask]):
+            print(token_counts, token_remainders)
```

Comment:
is this print for debugging purposes or you actually wanted to present it to a user?